### PR TITLE
(feature) Better Stash

### PR DESCRIPTION
**Pull request by nickisnoobatcode**
Added the better stash feature (originally made by itqxic in chat triggers)

image
image

### CODE REVIEW

**Comment by My-Name-Is-Jeff** [comment]
thanske asy nick

File: src/main/kotlin/skytils/skytilsmod/Skytils.kt [hunk 1]
```diff
@@ -276,6 +276,7 @@ class Skytils {
             TriviaSolver(),
             WaterBoardSolver(),
             Waypoints(),
+            BetterStash(),
```
Comment (My-Name-Is-Jeff) [inline src/main/kotlin/skytils/skytilsmod/Skytils.kt]: its alphabetical

File: src/main/kotlin/skytils/skytilsmod/core/Config.kt [hunk 2]
```diff
@@ -1896,6 +1896,13 @@ object Config : Vigilant(
     )
     var rareRelicFinder = false
 
+    @Property(
```
Comment (My-Name-Is-Jeff) [inline src/main/kotlin/skytils/skytilsmod/core/Config.kt]: why did you pick an arbitrary location

File: src/main/kotlin/skytils/skytilsmod/features/impl/misc/BetterStash.kt [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+
+class BetterStash {
+    var shouldDelete = false
+
+    @SubscribeEvent(priority = EventPriority.LOWEST)
+    fun onChat(event: ClientChatReceivedEvent) {
+        if (Skytils.config.betterStash && Utils.inSkyblock) {
+            if (Regex("§eYou have §a\\d+ §eitems? stashed away!!!\\n.*").matches(event.message.unformattedText)) {
+                event.isCanceled = true
+                Skytils.sendMessageQueue.add("/viewstash")
```
Comment (My-Name-Is-Jeff) [inline src/main/kotlin/skytils/skytilsmod/features/impl/misc/BetterStash.kt:46]: maybe add like a warning

File: src/main/kotlin/skytils/skytilsmod/features/impl/misc/BetterStash.kt [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+import skytils.skytilsmod.Skytils
+import skytils.skytilsmod.utils.Utils
+
+class BetterStash {
+    var shouldDelete = false
+
+    @SubscribeEvent(priority = EventPriority.LOWEST)
+    fun onChat(event: ClientChatReceivedEvent) {
+        if (Skytils.config.betterStash && Utils.inSkyblock) {
+            if (Regex("§eYou have §a\\d+ §eitems? stashed away!!!\\n.*").matches(event.message.unformattedText)) {
```
Comment (My-Name-Is-Jeff) [inline src/main/kotlin/skytils/skytilsmod/features/impl/misc/BetterStash.kt]: stop compiling the regex over and over.. this is is how sbe gets the best fps possible

File: src/main/kotlin/skytils/skytilsmod/features/impl/misc/BetterStash.kt [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+                event.isCanceled = true
+                Skytils.sendMessageQueue.add("/viewstash")
+                shouldDelete = true
+            } else if (shouldDelete && Regex("\\n§b§lItem Stash Contents:[\\s\\S]*").matches(event.message.unformattedText)) {
+                shouldDelete = false
+                event.isCanceled = true
+                val items =
+                    Regex("Item Stash Contents:\n([\\s\\S]*)").find(event.message.unformattedText)!!.groupValues[1].split("\n")
```
Comment (My-Name-Is-Jeff) [inline src/main/kotlin/skytils/skytilsmod/features/impl/misc/BetterStash.kt]: stop compiling the regex over and over.. this is is how sbe gets the best fps possible

File: src/main/kotlin/skytils/skytilsmod/features/impl/misc/BetterStash.kt [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+                        .filter {
+                            Regex("^§a\\d+(,\\d+)*x\\x20.*\$").matches(it)
```
Comment (My-Name-Is-Jeff) [inline src/main/kotlin/skytils/skytilsmod/features/impl/misc/BetterStash.kt]: stop compiling the regex over and over.. this is is how sbe gets the best fps possible

File: src/main/kotlin/skytils/skytilsmod/features/impl/misc/BetterStash.kt [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+                        }
+                Skytils.mc.thePlayer.addChatMessage(
```
Comment (My-Name-Is-Jeff) [inline src/main/kotlin/skytils/skytilsmod/features/impl/misc/BetterStash.kt]: Use UChat#chat

File: src/main/kotlin/skytils/skytilsmod/features/impl/misc/BetterStash.kt [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+                    ChatComponentText("§e§lYou have §a${items.size}x §eitems in your stash!")
+                )
+                Skytils.mc.thePlayer.addChatMessage(
```
Comment (My-Name-Is-Jeff) [inline src/main/kotlin/skytils/skytilsmod/features/impl/misc/BetterStash.kt]: Use UCraft methods

**Review by My-Name-Is-Jeff — CHANGES_REQUESTED** [review]

File: src/main/kotlin/skytils/skytilsmod/features/impl/misc/BetterStash.kt (hunk 3 above)
Comment (nickisnoobatcode) [reply src/main/kotlin/skytils/skytilsmod/features/impl/misc/BetterStash.kt:46]: wdym

File: src/main/kotlin/skytils/skytilsmod/core/Config.kt (hunk 2 above)
Comment (nickisnoobatcode) [reply src/main/kotlin/skytils/skytilsmod/core/Config.kt]: i put it with the other misc features

Comment (My-Name-Is-Jeff) [reply src/main/kotlin/skytils/skytilsmod/core/Config.kt]: you put it in between a feature and a button that goes with the feature

File: src/main/kotlin/skytils/skytilsmod/features/impl/misc/BetterStash.kt (hunk 3 above)
Comment (My-Name-Is-Jeff) [reply src/main/kotlin/skytils/skytilsmod/features/impl/misc/BetterStash.kt:46]: idk cause it runs commands just put like a thing in config description

File: src/main/kotlin/skytils/skytilsmod/features/impl/misc/BetterStash.kt [hunk 9]
```diff
@@ -0,0 +1,79 @@
...
+            } else if (shouldDelete && stashContentRegex.matches(event.message.unformattedText)) {
+                shouldDelete = false
+                event.isCanceled = true
+                val items = stashContentRegex2.find(event.message.unformattedText)!!.groupValues[1].split("\n")
+                    .filter {
+                        stashedItemRegex.matches(it)
+                    }
+
+                UChat.chat(
+                    UTextComponent("§e§lYou have §a${items.size}x §eitems in your stash!")
```
Comment (My-Name-Is-Jeff) [inline src/main/kotlin/skytils/skytilsmod/features/impl/misc/BetterStash.kt]: remove useless constructed object

File: src/main/kotlin/skytils/skytilsmod/features/impl/misc/BetterStash.kt [hunk 10]
```diff
@@ -0,0 +1,79 @@
...
+                )
+
+                UMessage(
+                    UTextComponent("§b§l[Items]").setClick(
+                        ClickEvent.Action.RUN_COMMAND, "/viewstash"
```
Comment (My-Name-Is-Jeff) [inline src/main/kotlin/skytils/skytilsmod/features/impl/misc/BetterStash.kt]: use type aliases MCHoverEventAction and MCClickEventAction

**Review by My-Name-Is-Jeff — CHANGES_REQUESTED** [review]

File: src/main/kotlin/skytils/skytilsmod/core/Config.kt [hunk 11]
```diff
@@ -1906,6 +1906,13 @@ object Config : Vigilant(
         Tracker.getTrackerById("found_spiders_den_relics")!!.doReset()
     }
 
+    @Property(
```
Comment (My-Name-Is-Jeff) [inline src/main/kotlin/skytils/skytilsmod/core/Config.kt]: can you move it to be in alphabetical order

File: src/main/kotlin/skytils/skytilsmod/core/Config.kt [hunk 12]
```diff
@@ -1906,6 +1906,13 @@ object Config : Vigilant(
...
+        type = PropertyType.SWITCH, name = "Better Stash",
+        description = "Shows an improved view of your stash.\n&cWARNING: This feature runs /viewstash upon receiving a stash message.",
```
Comment (My-Name-Is-Jeff) [inline src/main/kotlin/skytils/skytilsmod/core/Config.kt]: idk if & works does it

**Review by My-Name-Is-Jeff — APPROVED** [review]

File: src/main/kotlin/skytils/skytilsmod/core/Config.kt (hunk 12 above)
Comment (nickisnoobatcode) [reply src/main/kotlin/skytils/skytilsmod/core/Config.kt]: it does

**Pull request merged**